firecracker-control/local.go

```diff
@@ -81,263 +67,246 @@ func newLocal(ic *plugin.InitContext) (*local, error) {
 	}
 
 	return &local{
-		vm:          make(map[string]*instance),
-		rootPath:    ic.Root,
-		publisher:   ic.Events,
-		findVsockFn: findNextAvailableVsockCID,
+		containerdAddress: ic.Address,
+		logger:            log.G(ic.Context),
 	}, nil
 }
 
 // CreateVM creates new Firecracker VM instance
-func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*empty.Empty, error) {
+func (s *local) CreateVM(requestCtx context.Context, req *proto.CreateVMRequest) (*empty.Empty, error) {
+	var err error
+
 	id := req.GetVMID()
 	if id == "" {
-		return nil, errors.New("invalid VM ID")
+		err = errors.New("invalid VM ID")
+		s.logger.WithError(err).Error()
+		return nil, err
 	}
 
-	s.vmLock.Lock()
-
-	// Make sure VM ID unique
-	if _, ok := s.vm[id]; ok {
-		s.vmLock.Unlock()
-		return nil, errors.Errorf("VM with id %q already exists", id)
+	ns, err := namespaces.NamespaceRequired(requestCtx)
+	if err != nil {
+		err = errors.Wrap(err, "error retrieving namespace of request")
+		s.logger.WithError(err).Error()
+		return nil, err
 	}
 
-	// Reserve this ID while we're creating an instance in order to avoid race conditions
-	s.vm[id] = &instance{}
-	s.vmLock.Unlock()
-
-	instance, err := s.newInstance(ctx, id, req)
-
-	s.vmLock.Lock()
-	defer s.vmLock.Unlock()
+	s.logger.Debugf("using namespace: %s", ns)
 
+	// We determine if there is already a shim managing a VM with the current VMID by attempting
+	// to listen on the abstract socket address (which is parameterized by VMID). If we get
+	// EADDRINUSE, then we assume there is already a shim for the VM and just hand off to that
+	// one. This is in line with the approach used by containerd's reference runC shim v2 implementation.
+	shimSocketAddress, err := fcShim.SocketAddress(requestCtx, id)
 	if err != nil {
-		log.G(ctx).WithError(err).WithField("vm_id", id).Error("failed to create VM")
-		delete(s.vm, id)
-
+		err = errors.Wrap(err, "failed to obtain shim socket address")
+		s.logger.WithError(err).Error()
 		return nil, err
 	}
 
-	s.vm[id] = instance
-	return &empty.Empty{}, nil
-}
+	shimSocket, err := shim.NewSocket(shimSocketAddress)
+	if isEADDRINUSE(err) {
+		return nil, status.Errorf(codes.AlreadyExists, "VM with ID %q already exists", id)
+	} else if err != nil {
+		err = errors.Wrapf(err, "failed to open shim socket at address %q", shimSocketAddress)
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
 
-func (s *local) newInstance(ctx context.Context, id string, req *proto.CreateVMRequest) (*instance, error) {
-	logger := log.G(ctx).WithField("vm_id", id)
+	// If we're here, there is no pre-existing shim for this VMID, so we spawn a new one
```

Comment:
If we receive two requests simultaneously with the same ID, the code below will attempt to create a VM twice?

Reply:
If we receive two requests with the same ID one will get to this line and another will follow the `isEADDRINUSE` branch. Opening the `shimSocket` works essentially like a lock; only one process on the box can have the shimSocket for a given vmID open at a given time, anyone else that tries to open it gets `EADDRINUSE`.